### PR TITLE
fix(kimi): reconcile zero ratio placeholders with matching legacy counts

### LINKED ISSUE #3754: Kimi international Allegro: zero ratio fields override nonzero usage counts, showing 100% remaining
## Problem

On CodexBar 0.61.0, a Kimi **Allegro subscription on the international `kimi.ai` service** displays **100% remaining** for both the 5-hour and 7-day windows even though the subscription website shows nonzero usage. API-key authentication succeeds; this is a quota interpretation problem, not an authentication failure.

Observed together:

| Window | kimi.ai subscription page (used) | Code API count fields (used) | CodexBar (remaining) |
| --- | ---: | ---: | ---: |
| 5 hours | 1.02% | 1 / 100 | 100% |
| 7 days | 18.69% | 19 / 100 | 100% |

The count fields agree with the website after integer rounding. Expected API-derived remaining values are **99%** and **81%**. Monthly total usage is a separate metric and is not the subject of this report.

## Environment and reproduction

- CodexBar 0.61.0, macOS, Kimi Code provider
- Allegro subscription on `https://www.kimi.ai/settings/subscription`
- Kimi Code API key; international endpoint `https://api.kimi.ai/coding/v1/usages`
- Reproduced with `KIMI_CODE_BASE_URL=https://api.kimi.ai/coding/v1 codexbar usage --provider kimi --source api --json`

1. Configure the Kimi Code API key and international endpoint.
2. Fetch usage for an account returning the mixed response below.
3. Compare CodexBar with the subscription page, or parse the sanitized fixture offline.

## Sanitized API response

No credentials, account identifiers, or request headers are included.

```json
{
  "usage": {
    "limit": "100",
    "used": "19",
    "remaining": "81",
    "resetTime": "2026-09-19T16:45:59.449979Z"
  },
  "limits": [{
    "window": {"duration": 300, "timeUnit": "TIME_UNIT_MINUTE"},
    "detail": {
      "limit": "100",
      "used": "1",
      "remaining": "99",
      "resetTime": "2026-09-19T14:45:59.449979Z"
    }
  }],
  "usages": {
    "limit_5h": {"used_ratio": 0, "reset_time": "2026-09-19T14:45:58Z"},
    "limit_7d": {"used_ratio": 0, "reset_time": "2026-09-19T16:45:58Z"}
  }
}
```

CodexBar normalizes this to `primary.usedPercent = 0` and `secondary.usedPercent = 0`. `KimiUsageSnapshot.toUsageSnapshot()` unconditionally prefers valid ratio windows over count windows, including these conflicting zeros.

## Scope of a fix

Recognize the mixed legacy-count response without broadly changing ratio-pool precedence for other accounts. Preserve nonzero ratios, monthly-pool responses, missing/invalid-counter behavior, and genuine reset transitions. The observed reset timestamps differ by approximately 1.45 seconds; do not treat counters from a different quota period as current.

## Related work

- #3694 / #3697 added ratio-pool decoding and fixed a different missing-data failure. This report concerns an accepted response with conflicting values.
- #3536 / #3537 / #3543 concern missing or exhausted monthly membership usage, not this nonzero short-window mismatch.
- #3752 concerns the international dashboard destination; changing that link does not resolve this parser behavior.

A narrowly scoped follow-up PR with offline regression coverage will reference this issue.

### PR DESCRIPTION
**Pull request by mudrii**
## Problem and behavior

Closes #3754.

A mixed Kimi Code response for an international Allegro account contains nonzero legacy counts (`19/100` weekly and `1/100` over five hours) alongside zero ratio fields for the same windows. CodexBar currently prioritizes the zeros and displays **100% remaining** for both. The website shows 18.69% and 1.02% used, consistent with the rounded counts.

This change uses the populated counters in that bounded mixed-response case, yielding **81% weekly and 99% five-hour remaining**.

## Scope

Fall back from a zero ratio only when:

- A reliable legacy weekly count exists and no monthly ratio pool is present.
- The corresponding count is reliable and nonzero.
- The count and ratio windows have the same duration and reset timestamps within two seconds. The reported response has approximately 1.45 seconds of timestamp skew.

Nonzero ratios, monthly-pool responses, missing or invalid counters, and different reset periods retain their existing behavior. This preserves the ratio-only response support introduced by #3697. Authentication, endpoint selection, and monthly quota selection are unchanged.

The sanitized fixture contains no credentials or account identifiers. Regression coverage includes both affected windows, reset precision/skew, nonzero ratio precedence, monthly pools, invalid counters/resets, and mismatched durations. Provider documentation and the Unreleased changelog explain the bounded fallback.

## Validation

- The new quota regression assertions fail against unchanged upstream `KimiUsageSnapshot.swift`.
- **15 tests in `KimiRatioPoolTests` pass** against the corrected core. Initially, these were run offline in a temporary core-only SwiftPM harness using this exact test file and the repository's `CodexBarCore` library. No production source or test behavior was substituted, and those focused tests did not access live credentials or the browser. A separate, explicitly authorized live check is recorded below.
- **`make check` passes**, including SwiftFormat and SwiftLint (zero violations across 2,444 Swift files). The local invocation selected the installed Python 3.13 and Command Line Tools SourceKit explicitly.
- `git diff --check` passes.
- **Full `make test` now passes with configured Xcode** on head `916311b`: 1,233 discovered selections, all 113 groups successful on the first attempt, zero failed groups, retries, or timeouts. Total duration: 1,162 seconds. Used `DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer`, Python 3.13, and `CODEXBAR_SUPPRESS_TEST_KEYCHAIN_ACCESS=1`. Upstream CI approval remains separate.

### Local focused-test workaround

The temporary test package depends on this checkout's `CodexBarCore` product and copies `Tests/CodexBarTests/KimiRatioPoolTests.swift` unchanged. Its test invocation was:

```sh
CODEXBAR_SUPPRESS_TEST_KEYCHAIN_ACCESS=1 swift test \
  --package-path ../kimi-regression --scratch-path .build \
  --filter KimiRatioPoolTests \
  -Xswiftc -load-plugin-library \
  -Xswiftc /Library/Developer/CommandLineTools/usr/lib/swift/host/plugins/testing/libTestingMacros.dylib
```

With full Xcode configured, the normal repository validation is `swift test --filter KimiRatioPoolTests`, followed by `make test` and `make check`.

## Live affected-account verification

Verified the clean PR head `916311bdaa832224ec460a54c30ffff410fc20e7` on 2026-09-19 at 15:59 UTC against the affected international Allegro account using `https://api.kimi.ai/coding/v1/usages`. Usage has advanced since the original reproduction.

| Window | Refreshed kimi.ai website: used | Live API counts: used | Installed 0.61.0 CLI: used | Patched CLI: used |
| --- | ---: | ---: | ---: | ---: |
| Five hours | 3.62% | 4/100 | 0% | 4% |
| Seven days | 19.59% | 20/100 | 0% | 20% |

The patched values match the API counters and the website after rounding. The website provides finer precision than these counters. This is an actual live account check, separate from the offline regression fixture.

Built the production CLI with `swift build --product CodexBarCLI`. Ran both the installed helper and the built `.build/debug/CodexBarCLI` with `usage --provider kimi --source api --json`, selecting the international API endpoint through `KIMI_CODE_BASE_URL=https://api.kimi.ai/coding/v1`. The existing locally configured key was kept private. The installed GUI application was not replaced.

Patched binary SHA-256: `ef7125ff8c93414efd61d8963ff971aa92677af8b3d9921e00a270b542dc17dc`.

Sanitized projection of the live API response and actual CLI JSON outputs (only quota fields retained; no credentials or account identifiers):

```json
{
  "api_response": {
    "usage": {
      "limit": "100",
      "used": "20",
      "remaining": "80",
      "resetTime": "2026-09-19T16:45:59.449979Z"
    },
    "limits": [
      {
        "window": {
          "duration": 300,
          "timeUnit": "TIME_UNIT_MINUTE"
        },
        "detail": {
          "limit": "100",
          "used": "4",
          "remaining": "96",
          "resetTime": "2026-09-19T19:45:59.449979Z"
        }
      }
    ],
    "usages": {
      "limit_5h": {
        "used_ratio": 0,
        "reset_time": "2026-09-19T19:45:59Z"
      },
      "limit_7d": {
        "used_ratio": 0,
        "reset_time": "2026-09-19T16:45:59Z"
      }
    }
  },
  "installed_0_61_0": {
    "source": "Kimi Code API key",
    "weekly": {
      "usedPercent": 0,
      "windowMinutes": 10080,
      "resetsAt": "2026-09-19T16:45:58Z"
    },
    "five_hour": {
      "usedPercent": 0,
      "windowMinutes": 300,
      "resetsAt": "2026-09-19T19:45:58Z"
    }
  },
  "patched": {
    "source": "Kimi Code API key",
    "weekly": {
      "usedPercent": 20,
      "windowMinutes": 10080,
      "resetsAt": "2026-09-19T16:45:59Z",
      "resetDescription": "20/100 requests"
    },
    "five_hour": {
      "usedPercent": 4,
      "windowMinutes": 300,
      "resetsAt": "2026-09-19T19:45:59Z",
      "resetDescription": "Rate: 4/100 per 5 hours"
    }
  }
}
```

Both patched percentages were programmatically checked against the live API count percentages.

## Review follow-up

- Moved the changelog entry into the existing versioned Unreleased section.
- Used a tolerant floating-point comparison in the nonzero-ratio regression test.
- Re-ran all 15 focused tests, `make check`, the production CLI build, and the live check on the updated commit. Full `make test` subsequently passed after Xcode setup was completed: all 113 groups succeeded with no retries.

## Local packaged application validation

After completing Xcode setup, built the same clean head with the repository's `Scripts/package_app.sh release` using local ad-hoc signing. Packaging succeeded, including the widget extension, strict code-signature verification, bundled and symlinked helper resource probes, and a six-second standalone app launch smoke test with the checkout unavailable.

Copied the bundle outside the checkout and launched that exact copy. Confirmed its process path and strict signature verification. Its bundled Kimi CLI successfully fetched current quota data. The installed production application on disk was preserved.

The full repository suite also passed: 1,233 selected test selections across 113 groups, all successful on first pass; zero retries, failures, or timeouts.

## Maintainer validation on current main

Integrated current main and moved the release note into 0.62.1 Unreleased with contributor credit. Final integrated head: `b0ef7ba0eba5e990d638efda2d0d7712b0869fdb`.

- `CODEXBAR_SUPPRESS_TEST_KEYCHAIN_ACCESS=1 CODEXBAR_ALLOW_TEST_KEYCHAIN_ACCESS=0 make test-fast FILTER='KimiRatioPoolTests|KimiProviderTests'`: 88 tests in 7 suites passed.
- `CODEXBAR_SUPPRESS_TEST_KEYCHAIN_ACCESS=1 CODEXBAR_ALLOW_TEST_KEYCHAIN_ACCESS=0 make test`: all 1,243 selections across 114 groups passed. Two unrelated groups recovered on their automatic retry: the filesystem watcher notification test and a background-refresh coalescing test. No timeouts; total 1,348 seconds.
- `make check`: passed, including SwiftFormat and SwiftLint (zero violations across 2,458 Swift files).
- Independent Codex AutoReview of the complete integrated diff: no actionable P0–P2 findings.
- Both earlier review conversations are resolved. Maintainer validation used isolated tests, without live credential probes. The contributor's affected-account comparison above supplies the live response evidence.

### CODE REVIEW

File: Tests/CodexBarTests/KimiRatioPoolTests.swift [hunk 1]
```diff
@@ -127,6 +127,122 @@ struct KimiRatioPoolTests {
...
+        #expect(usage.extraRateWindows == nil)
+    }
+
+    @Test(arguments: ["0.1869", "0.5"])
+    func `nonzero ratios remain authoritative over legacy counts`(_ ratio: String) throws {
+        let usage = try Self.parse("""
+        {"usage":{"limit":"100","used":"19","resetTime":"2026-09-19T16:45:59Z"},
+         "usages":{"limit_7d":{"used_ratio":\(ratio),"reset_time":"2026-09-19T16:45:59Z"}}}
+        """)
+        #expect(usage.primary?.usedPercent == Double(ratio)! * 100)
```
[the remark this answers is not in the document]
Comment (mudrii) [reply Tests/CodexBarTests/KimiRatioPoolTests.swift]: Fixed in 916311b: the nonzero-ratio assertion now uses an absolute-difference tolerance. All 15 focused Kimi ratio tests pass.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## Unreleased
+
+### Fixed
+
+- Kimi: retain nonzero weekly and five-hour counts when a mixed legacy response includes conflicting zero ratios for the same quota windows (#3754).
+
 ## 0.61.1 — Unreleased
```
[the remark this answers is not in the document]
Comment (mudrii) [reply CHANGELOG.md]: Fixed in 916311b: removed the duplicate generic Unreleased section and moved the Kimi entry into the existing 0.61.1 Unreleased Fixed section.

**Comment by mudrii** [comment]
@clawsweeper re-review

Updated head: 916311bdaa832224ec460a54c30ffff410fc20e7. The PR description now contains sanitized live affected-account evidence from this exact clean commit, including the production CLI binary hash and raw quota projections. Installed 0.61.0 reports 0%/0%; the patched CLI reports 4% five-hour and 20% weekly, matching the live API counts and rounded kimi.ai website readings. Both Copilot comments are fixed; 15 focused tests and make check pass.

A maintainer may need to approve the fork workflow before full macOS CI can run; local full-suite execution remains blocked by Xcode setup.

**Comment by mudrii** [comment]
@clawsweeper re-review

The remaining full-suite evidence is now supplied: make test passes with configured Xcode on unchanged head 916311b, all 113 groups successful on the first attempt (1,233 selections), zero failures/retries/timeouts. The release app and widget also package successfully, pass signature/resource/standalone launch smoke checks, and the exact copied test app is running locally. The PR body records the details. Upstream fork CI approval remains a separate maintainer gate.

**Comment by steipete** [comment]
Landed as fd2414d295767d263f768848c540870b95eb918f. Thanks @mudrii for the precise reproduction, regression coverage, and affected-account comparison.

Validated integrated head b0ef7ba0eba5e990d638efda2d0d7712b0869fdb against current main:

- `CODEXBAR_SUPPRESS_TEST_KEYCHAIN_ACCESS=1 CODEXBAR_ALLOW_TEST_KEYCHAIN_ACCESS=0 make test-fast FILTER='KimiRatioPoolTests|KimiProviderTests'`: 88 tests in 7 suites passed.
- `CODEXBAR_SUPPRESS_TEST_KEYCHAIN_ACCESS=1 CODEXBAR_ALLOW_TEST_KEYCHAIN_ACCESS=0 make test`: all 1,243 selections across 114 groups passed in 1,348 seconds. Two unrelated timing-sensitive groups passed on automatic retry (filesystem watcher and background-refresh coalescing); no timeouts.
- `make check`: passed, including SwiftFormat and SwiftLint with zero violations across 2,458 Swift files.
- Independent Codex AutoReview: no actionable P0–P2 findings in the complete integrated diff.
- [CI run 35472195888](https://github.com/steipete/CodexBar/actions/runs/35472195888): green for both macOS test shards, Linux x64, Linux arm64, Linux musl, lint, and aggregate checks.

The contributor's sanitized live comparison supplies the affected-account proof; maintainer validation used isolated tests and no live credential probes. The shared converter now produces 19% weekly and 1% five-hour usage for the reported fixture, while preserving the existing monthly and reset-period behavior. The release note is under 0.62.1 Unreleased; this merge does not publish a release.
